torch/distributed/_composable/fsdp/fully_shard.py

```diff
@@ -163,6 +163,39 @@ def set_is_last_backward(self, is_last_backward: bool) -> None:
         state = self._get_fsdp_state()
         state._state_ctx.is_last_backward = is_last_backward
 
+    def set_requires_gradient_sync(
+        self, requires_gradient_sync: bool, recurse: bool = True
+    ) -> None:
+        """
+        Sets if the module should sync gradients. This can be used to implement
```

Comment:
hmmm I wonder if we should allow user to totally turn off grad sync without turn the grad sync back on later?

FSDP normally only have two modes: 1. fully shard 2. HSDP, for either case, we need to do reduce_scatter and all_reduce. But if user set the `requires_gradient_sync=False` and forget to reduce the grads, then FSDP won't do anything for the model grads, wouldn't that make DP semantic be wrong?

Making user control whether to sync grad or not maybe necessary to compose with PP but I wonder if we should make this sth like a context manager and restore reduce_grad somehow later

Reply:
My reasoning for a setter instead of a context is as follows:
1. Whether a setter or a context, the usage is conventionally a function of `is_last_microbatch`, which is computed from the microbatch index. There should not be a risk of forgetting to re-enable this gradient sync.
    - The expected usage of the setter is `module.set_requires_gradient_sync(is_last_microbatch)`, not directly as `set_requires_gradient_sync(True)` or `set_requires_gradient_sync(False)`.
    - The current usage of `no_sync()` looks like `with root_module.no_sync() if not is_last_microbatch else contextlib.nullcontext()`.
2. Using a context manager makes it more clumsy to selectively disable gradient sync for some submodules but not others. The existing `no_sync()` must be called on the root FSDP module and applies to all child FSDP modules as well. However, we can imagine performance tuning where, when accumulating gradients across microbatches, we disable gradient sync for some submodules but not others to trade-off communication and memory.
3. `no_sync()` is not always intuitive. There have been issues where DDP users did not know that they must include both `forward()` and `backward()` under the context, leading to incorrect behavior (https://github.com/pytorch/pytorch/pull/89244). In contrast, FSDP does not need both `forward()` and `backward()` to be under the context; however, users may extrapolate from DDP. Karpathy gave this feedback in nanoGPT, where instead prefers setting a variable:
https://github.com/karpathy/nanoGPT/blob/eba36e84649f3c6d840a93092cb779a260544d08/train.py#L291-L295

Reply:
Ok yeah this convinced me!